markdown_docs/repo_agent/settings.md
## ClassDef LogLevel
**LogLevel**: Функция LogLevel представляет собой перечисление (enum) уровней логирования.

**Attributes**:
- DEBUG
- INFO
- WARNING
- ERROR
- CRITICAL

**Описание кода**: Класс LogLevel определяет уровни логирования, которые могут быть использованы для настройки уровня детализации выводимых сообщений. Эти уровни могут быть применены в настройках проекта для контроля за количеством и детализацией выводимых логов. Класс LogLevel используется в ProjectSettings для определения уровня логирования.

**Примечание**: При использовании класса LogLevel в настройках проекта, убедитесь, что значение уровня логирования соответствует одному из определённых в перечислении LogLevel. В противном случае будет выброшено исключение ValueError.
## ClassDef ProjectSettings
**ProjectSettings**: Функция ProjectSettings представляет собой класс, предназначенный для хранения настроек проекта.

**Attributes**:
* параметр 1: `target_repo` — путь к директории, где хранятся репозитории.
* параметр 2: `hierarchy_name` — имя иерархии для хранения документов.
* параметр 3: `markdown_docs_name` — имя для хранения markdown документов.
* параметр 4: `ignore_list` — список игнорируемых элементов.
* параметр 5: `language` — язык проекта (по умолчанию — English).
* параметр 6: `max_thread_count` — максимальное количество потоков для обработки (по умолчанию — 4).
* параметр 7: `log_level` — уровень логирования (по умолчанию — INFO).

**Описание кода**: Класс ProjectSettings предназначен для хранения и настройки параметров проекта. В классе определены атрибуты для настройки пути к репозиторию, имени иерархии, имени для хранения markdown документов, списка игнорируемых элементов, языка проекта, максимального количества потоков и уровня логирования.

Для проверки корректности вводимых значений языка и уровня логирования используются методы `validate_language_code` и `set_log_level`. Метод `validate_language_code` проверяет, соответствует ли введённое значение языка допустимому коду или имени языка, а метод `set_log_level` преобразует введённое значение уровня логирования в соответствующий элемент перечисления LogLevel.

Класс ProjectSettings используется в методе `initialize_with_params` класса SettingsManager для инициализации настроек проекта на основе переданных параметров.

**Примечание**: При использовании класса ProjectSettings убедитесь, что значения параметров соответствуют ожидаемому типу данных. В противном случае может быть выброшено исключение. Также обратите внимание на использование методов `validate_language_code` и `set_log_level` для корректной настройки параметров языка и уровня логирования.
### FunctionDef validate_language_code(cls)
**validate_language_code**: Функция validate_language_code используется для проверки корректности кода языка.

**parameters**:
* параметр 1: `v: str` — строка, содержащая код языка или название языка.

**Описание кода**:
Функция `validate_language_code` предназначена для валидации кода языка. Она принимает строку `v`, которая должна содержать код языка согласно стандарту ISO 639 или его название. Внутри функции используется класс `Language` для поиска соответствующего названия языка по переданному коду. Если код языка не найден, вызывается исключение `LanguageNotFoundError`, которое затем преобразуется в `ValueError` с сообщением об ошибке. В случае успешного поиска возвращается название языка.

**Примечание**:
При использовании функции `validate_language_code` необходимо убедиться, что передаваемый параметр `v` соответствует формату ISO 639 или является допустимым названием языка. В противном случае будет вызвано исключение `ValueError`.

**Пример вывода**:
Не применимо к данной функции.
***
### FunctionDef set_log_level(cls)
**set_log_level**: Функция `set_log_level` предназначена для установки уровня логирования в соответствии с переданным значением.

**parameters**:
- параметр 1: `v: str` — строка, представляющая уровень логирования.

**Описание кода**: Функция `set_log_level` проверяет и преобразует переданный уровень логирования в соответствии с перечислением `LogLevel`. Если переданное значение является строкой, оно приводится к верхнему регистру. Затем проверяется, соответствует ли преобразованное значение одному из определённых уровней логирования в перечислении `LogLevel`. Если значение соответствует, возвращается соответствующий объект `LogLevel`. В противном случае выбрасывается исключение `ValueError` с указанием некорректного уровня логирования.

**Примечание**: При использовании функции `set_log_level` убедитесь, что передаваемое значение уровня логирования соответствует одному из определённых в перечислении `LogLevel`. В противном случае будет выброшено исключение `ValueError`.
***
## FunctionDef get_openai_key
**get_openai_key**: Функция `get_openai_key` предназначена для получения ключа API OpenAI из файла конфигурации `config.yaml`.

**Параметры**: отсутствуют параметры.

**Описание кода**: функция `get_openai_key` открывает файл `config.yaml` в режиме чтения с использованием кодировки `utf-8`. Затем содержимое файла загружается в формате YAML с помощью `yaml.load`. Из загруженного словаря извлекается значение ключа `openai_key` из раздела `auth` и возвращается в виде строки.

Функция вызывается в классе `ChatCompletionSettings`, где полученное значение ключа используется для настройки параметра `openai_api_key`.

**Примечание**: убедитесь, что файл `config.yaml` существует в том же каталоге, где выполняется код, и содержит корректный ключ API OpenAI в нужном формате.

**Пример вывода**: `str` с ключом API OpenAI.
## ClassDef ChatCompletionSettings
**ChatCompletionSettings**: Класс ChatCompletionSettings предназначен для настройки параметров завершения чата.

**Attributes**:
* параметр 1: model — модель для завершения чата, по умолчанию «gpt-4o-mini».
* параметр 2: temperature — температура для генерации текста, по умолчанию 0.2.
* параметр 3: request_timeout — время ожидания ответа в секундах, по умолчанию 60.
* параметр 4: openai_base_url — базовый URL для API OpenAI, по умолчанию «https://api.openai.com/v1».
* параметр 5: openai_api_key — ключ API OpenAI, значение берётся из файла конфигурации config.yaml.

**Описание кода**: класс ChatCompletionSettings наследуется от класса BaseSettings и содержит параметры для настройки завершения чата. Значение ключа API OpenAI берётся из файла config.yaml с помощью функции get_openai_key. Эта функция открывает файл config.yaml в режиме чтения с использованием кодировки utf-8, загружает содержимое файла в формате YAML и извлекает значение ключа openai_key из раздела auth.

Класс ChatCompletionSettings используется в методе initialize_with_params класса SettingsManager для инициализации настроек завершения чата. Метод initialize_with_params принимает параметры target_repo, markdown_docs_name, hierarchy_name, ignore_list, language, max_thread_count, log_level, model, temperature, request_timeout, openai_base_url и создаёт экземпляр класса ChatCompletionSettings с соответствующими значениями параметров.

**Примечание**: убедитесь, что файл config.yaml существует в том же каталоге, где выполняется код, и содержит корректный ключ API OpenAI в нужном формате.
### FunctionDef convert_base_url_to_str(cls)
**convert_base_url_to_str**: Функция convert_base_url_to_str преобразует URL-адрес типа HttpUrl в строку.

**parameters**:
* параметр 1: openai_base_url — URL-адрес типа HttpUrl, который необходимо преобразовать в строку.

**Описание кода**:
Функция convert_base_url_to_str использует декоратор @field_validator для проверки значения параметра openai_base_url перед выполнением преобразования. Она также определена как класс метод (classmethod), что позволяет вызывать её без необходимости создания экземпляра класса. Функция принимает один параметр openai_base_url, который должен быть экземпляром типа HttpUrl, и возвращает строку, представляющую этот URL.

**Примечание**:
При использовании функции необходимо убедиться, что параметр openai_base_url является допустимым URL-адресом типа HttpUrl.

**Пример вывода**:
Функция возвращает строку, представляющую URL-адрес, например: "https://example.com".
***
## ClassDef Setting
**Setting**: Функция Setting представляет собой класс, предназначенный для хранения настроек проекта и параметров завершения чата.

**Attributes**:
* параметр 1: `project` — экземпляр класса `ProjectSettings`, содержащий настройки проекта.
* параметр 2: `chat_completion` — экземпляр класса `ChatCompletionSettings`, содержащий параметры завершения чата.

**Описание кода**: класс `Setting` наследуется от класса `BaseSettings` и содержит параметры для хранения настроек проекта и параметров завершения чата. В классе `Setting` определены атрибуты для хранения экземпляра класса `ProjectSettings` и экземпляра класса `ChatCompletionSettings`.

Класс `Setting` используется в методе `initialize_with_params` класса `SettingsManager` для инициализации настроек проекта и завершения чата на основе переданных параметров. Метод `initialize_with_params` создаёт экземпляр класса `Setting` с соответствующими значениями параметров.

**Примечание**: при использовании класса `Setting` убедитесь, что значения параметров соответствуют ожидаемому типу данных. В противном случае может быть выброшено исключение.
## ClassDef SettingsManager
**SettingsManager**: Функция SettingsManager предназначена для управления настройками проекта.

**Attributes**:
* параметр _setting_instance: Optional[Setting] — приватный атрибут класса, инициализируется значением None.

**Описание кода**: SettingsManager — это класс, который управляет настройками проекта. Он содержит метод get_setting, который возвращает экземпляр класса Setting, если он ещё не был создан. Если экземпляр Setting уже существует, метод возвращает его.

Также класс содержит метод initialize_with_params, который используется для инициализации настроек проекта. Этот метод принимает несколько параметров, включая путь к репозиторию, имя файла с документацией в формате Markdown, имя иерархии, список игнорируемых элементов, язык, максимальное количество потоков, уровень логирования, модель, температуру, время ожидания запроса и базовый URL для OpenAI.

Метод initialize_with_params создаёт экземпляры классов ProjectSettings и ChatCompletionSettings, а затем инициализирует атрибут _setting_instance класса SettingsManager с помощью созданного экземпляра класса Setting.

**Примечание**: при использовании метода initialize_with_params убедитесь, что переданные параметры соответствуют ожидаемому типу данных. В противном случае может быть выброшено исключение.
### FunctionDef get_setting(cls)
**get_setting**: Функция `get_setting` возвращает экземпляр класса `Setting`, который содержит настройки проекта и параметры завершения чата.

**parameters**: параметров у функции нет.

**Описание кода**: метод `get_setting` класса `SettingsManager` используется для получения экземпляра класса `Setting`. Если экземпляр `Setting` ещё не был создан (`cls._setting_instance` равен `None`), метод создаёт новый экземпляр класса `Setting`. Затем возвращает этот экземпляр.

**Примечание**: при использовании метода `get_setting` не требуется передавать параметры. Убедитесь, что класс `Setting` используется корректно и значения параметров соответствуют ожидаемому типу данных, чтобы избежать возможных исключений.
***
### FunctionDef initialize_with_params(cls)
**initialize_with_params**: Функция `initialize_with_params` предназначена для инициализации настроек проекта и параметров завершения чата на основе переданных параметров.

**parameters**:
* параметр 1: `target_repo` — путь к директории, где хранятся репозитории.
* параметр 2: `markdown_docs_name` — имя для хранения markdown документов.
* параметр 3: `hierarchy_name` — имя иерархии для хранения документов.
* параметр 4: `ignore_list` — список игнорируемых элементов.
* параметр 5: `language` — язык проекта (по умолчанию — English).
* параметр 6: `max_thread_count` — максимальное количество потоков для обработки (по умолчанию — 4).
* параметр 7: `log_level` — уровень логирования (по умолчанию — INFO).
* параметр 8: `model` — модель для завершения чата, по умолчанию «gpt-4o-mini».
* параметр 9: `temperature` — температура для генерации текста, по умолчанию 0.2.
* параметр 10: `request_timeout` — время ожидания ответа в секундах, по умолчанию 60.
* параметр 11: `openai_base_url` — базовый URL для API OpenAI, по умолчанию «https://api.openai.com/v1».

**Описание кода**: функция `initialize_with_params` создаёт экземпляр класса `Setting` с соответствующими значениями параметров на основе переданных параметров. Для этого она сначала создаёт экземпляр класса `ProjectSettings` с переданными значениями параметров, а затем экземпляр класса `ChatCompletionSettings`. После этого она устанавливает значения атрибутов экземпляра класса `Setting` с помощью созданных экземпляров классов `ProjectSettings` и `ChatCompletionSettings`.

**Примечание**: при использовании функции `initialize_with_params` убедитесь, что значения параметров соответствуют ожидаемому типу данных. В противном случае может быть выброшено исключение. Также обратите внимание на использование методов `validate_language_code` и `set_log_level` для корректной настройки параметров языка и уровня логирования в классе `ProjectSettings`.
***


repo_agent/settings.py
from enum import StrEnum  # type: ignore
import yaml
from typing import Optional

from iso639 import Language, LanguageNotFoundError  # type: ignore
from pydantic import (
    DirectoryPath,
    Field,
    HttpUrl,
    PositiveFloat,
    PositiveInt,
    SecretStr,
    field_validator,
)
from pydantic_settings import BaseSettings  # type: ignore
from pathlib import Path

with open("config.yaml", "r", encoding="utf8") as f:
    config = yaml.load(f, Loader=yaml.FullLoader)


class LogLevel(StrEnum):
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARNING = "WARNING"
    ERROR = "ERROR"
    CRITICAL = "CRITICAL"


class ProjectSettings(BaseSettings):
    target_repo: DirectoryPath = ""  # type: ignore
    hierarchy_name: str = ".project_doc_record"
    markdown_docs_name: str = "markdown_docs"
    ignore_list: list[str] = []
    language: str = "English"
    max_thread_count: PositiveInt = 4
    log_level: LogLevel = LogLevel.INFO  # type: ignore

    @field_validator("language")
    @classmethod
    def validate_language_code(cls, v: str) -> str:
        try:
            language_name = Language.match(v).name
            return language_name  # Returning the resolved language name
        except LanguageNotFoundError:
            raise ValueError(
                "Invalid language input. Please enter a valid ISO 639 code or language name."
            )

    @field_validator("log_level", mode="before")
    @classmethod
    def set_log_level(cls, v: str) -> LogLevel:
        if isinstance(v, str):
            v = v.upper()  # Convert input to uppercase
        if (
            v in LogLevel._value2member_map_
        ):  # Check if the converted value is in enum members
            return LogLevel(v)
        raise ValueError(f"Invalid log level: {v}")


def get_openai_key() -> str:
    with open("config.yaml", "r", encoding="utf8") as f:
        config = yaml.load(f, Loader=yaml.FullLoader)
    return config["auth"]["openai_key"]


class ChatCompletionSettings(BaseSettings):
    model: str = "gpt-4o-mini"  # TODO
    temperature: PositiveFloat = 0.2
    request_timeout: PositiveInt = 60
    openai_base_url: str = "https://api.openai.com/v1"  # TODO
    openai_api_key: str = get_openai_key()

    @field_validator("openai_base_url", mode="before")
    @classmethod
    def convert_base_url_to_str(cls, openai_base_url: HttpUrl) -> str:
        return str(openai_base_url)


class Setting(BaseSettings):
    project: ProjectSettings = {}  # type: ignore
    chat_completion: ChatCompletionSettings = {}  # type: ignore


class SettingsManager:
    _setting_instance: Optional[Setting] = (
        None  # Private class attribute, initially None
    )

    @classmethod
    def get_setting(cls):
        if cls._setting_instance is None:
            cls._setting_instance = Setting()
        return cls._setting_instance

    @classmethod
    def initialize_with_params(
        cls,
        target_repo: Path,
        markdown_docs_name: str,
        hierarchy_name: str,
        ignore_list: list[str],
        language: str,
        max_thread_count: int,
        log_level: str,
        model: str,
        temperature: float,
        request_timeout: int,
        openai_base_url: str,
    ):
        project_settings = ProjectSettings(
            target_repo=target_repo,
            hierarchy_name=hierarchy_name,
            markdown_docs_name=markdown_docs_name,
            ignore_list=ignore_list,
            language=language,
            max_thread_count=max_thread_count,
            log_level=LogLevel(log_level),
        )

        chat_completion_settings = ChatCompletionSettings(
            model=model,
            temperature=temperature,
            request_timeout=request_timeout,
            openai_base_url=openai_base_url,
        )

        cls._setting_instance = Setting(
            project=project_settings,
            chat_completion=chat_completion_settings,
        )


if __name__ == "__main__":
    setting = SettingsManager.get_setting()
    print(setting.model_dump())  # type: ignore
